linux/devdoc/file_linux_requirements.md

```diff
@@ -0,0 +1,159 @@
+```
+
+**SRS_FILE_LINUX_43_034: [** If `handle` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_LINUX_43_035: [** If `user_callback` is `NULL` then `file_read_async` shall fail and return `FILE_READ_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_LINUX_43_016: [** `file_read_async` shall create a `FILE_LINUX_IO` struct with `handle` as `handle`, `FILE_ASYNC_READ` as `type`. **]**
+
+**SRS_FILE_LINUX_43_017: [** `file_read_async` shall populate the `data` field of the `FILE_LINUX_IO` struct using `user_callback` and `user_context`. **]**
+
+**SRS_FILE_LINUX_43_008: [** `file_read_async` shall initialize a `sigevent` struct with `SIGEV_THREAD` as `sigev_notify`, `user_context` as `sigev_value`, `user_callback` as `sigev_notify_function`, `NULL` as `sigev_notify_attributes`(unsure about what `sigev_notify_attributes` should be). **]**
+
+**SRS_FILE_LINUX_43_009: [** `file_read_async` shall initialize an `aiocb` struct with `handle` as `aio_fildes`, `position` as `aio_offset`, source as `aio_buf`, `source->alias->size` as `aio_nbytes`, and the initialized `sigevent` struct as `aio_sigevent`. **]**
+
+**SRS_FILE_LINUX_43_010: [** `file_read_async` shall call `aio_read` with the initialized `aiocb` struct as `aiocbp`.  **]**
+
+**SRS_FILE_LINUX_43_011: [** If `aio_read` fails, `file_read_async` shall return `FILE_READ_ASYNC_READ_ERROR`. **]**
+
+**SRS_FILE_LINUX_43_014: [** If `aio_read` succeeds, `file_read_async` shall return `FILE_READ_ASYNC_OK`. **]**
+
+**SRS_FILE_LINUX_43_015: [** If there are any failures, `file_read_async` shall return `FILE_READ_ASYNC_ERROR`. **]**
+
+
+## file_extend_filesize
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);
+```
+
+**SRS_FILE_LINUX_43_018: [** `file_extend_filesize` shall return 0. **]**
+
+Will be implemented later.
+
+## on_file_io_complete_linux
+
+```c
+static void on_file_io_complete_linux( FILE_LINUX_IO* io);
+```
+
+`on_file_io_complete_linux` is called when a file operation completes asynchronously.
+
+**SRS_FILE_LINUX_43_021: [** `on_file_io_complete_linux` shall recover the `aiocb` struct that was used to create the current asynchronous operation. **]**
+
+**SRS_FILE_LINUX_43_022: [** `on_file_io_complete_linux` shall call `aio_return` to determine if the asynchronous operation succeeded. **]**
+
+**SRS_FILE_LINUX_43_023: [** If the asynchronous operation did not succeed, `on_file_io_complete_linux` shall call `user_callback` with `false` as `is_successful`. **]**
+
+**SRS_FILE_LINUX_43_024: [** If the type of the asynchronous operation is read, `on_file_io_complete_linux` shall  construct a `CONSTBUFFER_HANDLE` by calling `CONSTBUFFER_CreateWithMoveMemory` from the bytes read by `ReadFile`. **]**

```

Comment:
>If the type of the asynchronous operation is read [](start = 29, length = 49)

how does it do this determination?

Reply:
FILE_LINUX_IO has a type field.

---
In reply to: [450478838](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450478838) [](ancestors = 450478838)